src/dataset/mod.rs

```diff
@@ -561,15 +561,25 @@ mod tests {
         }
     }
 
-    impl<'b> PredictRef<ArrayView2<'b, f64>, Array1<f64>> for MockFittableResult {
-        fn predict_ref<'a>(&'a self, _x: &'a ArrayView2<'b, f64>) -> Array1<f64> {
-            array![0.]
+    impl<'b> PredictInto<ArrayView2<'b, f64>, Array1<f64>> for MockFittableResult {
+        fn predict_into<'a>(&'a self, x: &'a ArrayView2<'b, f64>, y: &mut Array1<f64>) {
+            assert_eq!(
+                x.nrows(),
+                y.len(),
+                "The number of data points must match the number of output targets."
+            );
+            *y = array![0.];
         }
     }
 
-    impl<'b> PredictRef<ArrayView2<'b, f64>, Array2<f64>> for MockFittableResult {
-        fn predict_ref<'a>(&'a self, _x: &'a ArrayView2<'b, f64>) -> Array2<f64> {
-            array![[0., 0.]]
+    impl<'b> PredictInto<ArrayView2<'b, f64>, Array2<f64>> for MockFittableResult {
+        fn predict_into<'a>(&'a self, x: &'a ArrayView2<'b, f64>, y: &mut Array2<f64>) {
+            assert_eq!(
+                x.nrows(),
+                y.nrows(),
+                "The number of data points must match the number of output targets."
+            );
+            *y = array![[0., 0.]];
```

Comment:
as a note: your goal of efficient implementations is not achieved like this, you are making an allocation of array and just throwing the old one away. For example here, you would have to call `y.fill(0.0)` here. The infrastructure laid out here is just the first step, each implementation I've seen in this PR have to be modified to really benefit in terms of memory usage

Reply:
Yes, what I was doing is first replacing the `PredictRef`, and then actually implementing the more efficient versions. (The only one that I did implement a more efficient version is KMeans for now)

Reply:
ah I see, this is the way to go :rocket: